src/contract.rs

```diff
@@ -196,39 +195,42 @@ pub fn authenticated_queries<S: Storage, A: Api, Q: Querier>(
     deps: &Extern<S, A, Q>,
     msg: QueryMsg,
 ) -> QueryResult {
-    let (address, key) = msg.get_validation_params();
+    let (addresses, key) = msg.get_validation_params();
 
-    let canonical_addr = deps.api.canonical_address(address)?;
+    for address in addresses {
+        let canonical_addr = deps.api.canonical_address(address)?;
 
-    let expected_key = read_viewing_key(&deps.storage, &canonical_addr);
+        let expected_key = read_viewing_key(&deps.storage, &canonical_addr);
 
-    if expected_key.is_none() {
-        // Checking the key will take significant time. We don't want to exit immediately if it isn't set
-        // in a way which will allow to time the command and determine if a viewing key doesn't exist
-        key.check_viewing_key(&[0u8; 24]);
-        return Ok(to_binary(&QueryAnswer::ViewingKeyError {
-            msg: "Wrong viewing key for this address or viewing key not set".to_string(),
-        })?);
-    }
+        if expected_key.is_none() {
+            // Checking the key will take significant time. We don't want to exit immediately if it isn't set
+            // in a way which will allow to time the command and determine if a viewing key doesn't exist
+            key.check_viewing_key(&[0u8; VIEWING_KEY_SIZE]);
+            continue;
+        }
 
-    if !key.check_viewing_key(expected_key.unwrap().as_slice()) {
-        return Ok(to_binary(&QueryAnswer::ViewingKeyError {
-            msg: "Wrong viewing key for this address or viewing key not set".to_string(),
-        })?);
+        if key.check_viewing_key(expected_key.unwrap().as_slice()) {
+            return match msg {
+                // Base
+                QueryMsg::Balance { address, .. } => query_balance(&deps, &address),
+                QueryMsg::TransferHistory {
+                    address,
+                    page,
+                    page_size,
+                    ..
+                } => query_transactions(&deps, &address, page.unwrap_or(0), page_size),
+                QueryMsg::Allowance { owner, spender, .. } => {
+                    try_check_allowance(deps, owner, spender)
+                }
+                // Other - Test
+                _ => unimplemented!(),
+            };
+        }
     }
```

Comment:
Instead of returning from inside the loop, you could have a boolean variable `authenticated` that starts as `false`, and switches to `true` when a match is found. then later if it's still false, then return before doing the main `match`.
Your way is more efficient, but i think this will be easier to follow

Reply:
Based on the fact that the viewing key validation is _kinda_ expensive, I think it's better this way. This function is not too complex

Reply:
fair. it's cool this way too.